man/man3/Draw.3
.TH "Draw" 3 "Thu Jan 15 2015" "Version 9.9" "Version" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Draw \- Die \fBDraw\fP Klasse (ruft die Klassen \fBCharacter\fP, \fBBackground\fP und UiOverlay auf)\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <draw\&.h>\fP
.PP
Inherits QWidget\&.
.SS "Public Slots"

.in +1c
.ti -1c
.RI "int \fBfps\fP ()"
.br
.RI "\fIWird angetriggert wenn die fps angezeigt wird\&. \fP"
.ti -1c
.RI "void \fBwinC\fP ()"
.br
.RI "\fIWird angetriggert wenn der erste Spieler gewinnt\&. \fP"
.ti -1c
.RI "void \fBwinE\fP ()"
.br
.RI "\fIWird angetriggert wenn der zweite Spieler/KI gewinnt\&. \fP"
.ti -1c
.RI "void \fBkiActionTrue\fP ()"
.br
.RI "\fIWird angetriggert wenn die KI wieder einen Zug machen soll\&. \fP"
.in -1c
.SS "Signals"

.in +1c
.ti -1c
.RI "void \fBchatSignal\fP ()"
.br
.RI "\fISignalisiert dem \fBChat\fP das er versteckt oder angezeigt werden soll\&. \fP"
.ti -1c
.RI "void \fBshowOnlyChat\fP ()"
.br
.RI "\fIZeigt nur den \fBChat\fP\&. \fP"
.ti -1c
.RI "void \fBhideOnlyChat\fP ()"
.br
.RI "\fIVersteckt nur den \fBChat\fP\&. \fP"
.ti -1c
.RI "void \fBsetCurrent\fP (int)"
.br
.RI "\fISignalisiert der Main das ein anderes Widget angezeigt werden soll\&. \fP"
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBDraw\fP (QWidget *parent=0)"
.br
.RI "\fIConstructor der Klasse \fBDraw\fP\&. \fP"
.ti -1c
.RI "void \fBsetFpsVisible\fP (bool b)"
.br
.RI "\fISetzt die Fps sichtbar\&. \fP"
.ti -1c
.RI "bool \fBisFpsVisible\fP ()"
.br
.RI "\fIGibt an ob die Fps sichtbar sind\&. \fP"
.ti -1c
.RI "void \fBload\fP (int selChar, int selBackg, int selChar2)"
.br
.RI "\fILadet die gewählten Charaktere und den Hintergrund\&. \fP"
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "void \fBpaintEvent\fP (QPaintEvent *e)"
.br
.RI "\fIZeichnet Elemente für die Klasse\&. \fP"
.ti -1c
.RI "void \fBkeyPressEvent\fP (QKeyEvent *e)"
.br
.RI "\fIRegistriert die Tastenschläge des Benutzers\&. \fP"
.ti -1c
.RI "void \fBkeyReleaseEvent\fP (QKeyEvent *e)"
.br
.RI "\fIRegistriert wenn der Benutzer die Taste loslässt\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
Die \fBDraw\fP Klasse (ruft die Klassen \fBCharacter\fP, \fBBackground\fP und UiOverlay auf)\&. 

Die Klasse \fBDraw\fP ruft die Klassen \fBCharacter\fP, \fBBackground\fP und UiOverlay auf\&. Diese sind für das Zeichnen im Spiel verantwortlich\&. \fBCharacter\fP und \fBBackground\fP werden Zahlen übergeben welche für die jeweilige Auswahl stehen (\fBBackground\fP 1, 2, \&.\&.\&. etc\&.)\&. Die UiOverlay Klasse brauch zum Zeichnen immer den \fBCharacter\fP als object, damit Healthbar etc\&. korrekte Werte zeichnen\&. 
.PP
Definition at line 21 of file draw\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Draw::Draw (QWidget * parent = \fC0\fP)\fC [explicit]\fP"

.PP
Constructor der Klasse \fBDraw\fP\&. Contructor der Klasse \fBDraw\fP\&. 
.PP
\fBParameters:\fP
.RS 4
\fIparent\fP Übergibt falls vorhanden das parent-QWidget\&. 
.RE
.PP

.PP
Definition at line 3 of file draw\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void Draw::chatSignal ()\fC [signal]\fP"

.PP
Signalisiert dem \fBChat\fP das er versteckt oder angezeigt werden soll\&. chatSignal 
.SS "void Draw::hideOnlyChat ()\fC [signal]\fP"

.PP
Versteckt nur den \fBChat\fP\&. hideOnlyChat 
.SS "bool Draw::isFpsVisible ()"

.PP
Gibt an ob die Fps sichtbar sind\&. isFpsVisible 
.PP
\fBReturns:\fP
.RS 4
showFps 
.RE
.PP

.PP
Definition at line 127 of file draw\&.cpp\&.
.SS "void Draw::keyPressEvent (QKeyEvent * e)\fC [protected]\fP"

.PP
Registriert die Tastenschläge des Benutzers\&. keyPressEvent 
.PP
\fBParameters:\fP
.RS 4
\fIe\fP Die Taste welche gedrückt wird\&. 
.RE
.PP

.PP
Definition at line 131 of file draw\&.cpp\&.
.SS "void Draw::keyReleaseEvent (QKeyEvent * e)\fC [protected]\fP"

.PP
Registriert wenn der Benutzer die Taste loslässt\&. keyReleaseEvent 
.PP
\fBParameters:\fP
.RS 4
\fIe\fP Die Taste welche losgelassen wurde\&. 
.RE
.PP

.PP
Definition at line 185 of file draw\&.cpp\&.
.SS "void Draw::kiActionTrue ()\fC [slot]\fP"

.PP
Wird angetriggert wenn die KI wieder einen Zug machen soll\&. \fBDraw::kiActionTrue\fP\&.
.PP
Setzt die KI Action true sodass die KI beim nächsten Zug wieder eine Aktion machen soll\&. 
.PP
Definition at line 283 of file draw\&.cpp\&.
.SS "void Draw::load (int selChar, int selBackg, int selChar2)"

.PP
Ladet die gewählten Charaktere und den Hintergrund\&. load 
.PP
\fBParameters:\fP
.RS 4
\fIselChar\fP der gewählte Charakter\&. 
.br
\fIselBackg\fP der gewählte Hintergrund\&. 
.br
\fIselChar2\fP der gewählte 2\&.Charakter\&. 
.RE
.PP

.PP
Definition at line 219 of file draw\&.cpp\&.
.SS "void Draw::paintEvent (QPaintEvent * e)\fC [protected]\fP"

.PP
Zeichnet Elemente für die Klasse\&. paintEvent 
.PP
\fBParameters:\fP
.RS 4
\fIe\fP Das paintEvent\&. 
.RE
.PP

.PP
Definition at line 36 of file draw\&.cpp\&.
.SS "void Draw::setCurrent (int)\fC [signal]\fP"

.PP
Signalisiert der Main das ein anderes Widget angezeigt werden soll\&. setCurrent 
.PP
\fBParameters:\fP
.RS 4
\fIint\fP Das Widget welches angezeigt werden soll\&. 
.RE
.PP

.SS "void Draw::setFpsVisible (bool b)"

.PP
Setzt die Fps sichtbar\&. setFpsVisible 
.PP
\fBParameters:\fP
.RS 4
\fIb\fP True - Fps ist sichtbar\&.
.br
False - Fps nicht sichtbar\&. 
.RE
.PP

.PP
Definition at line 123 of file draw\&.cpp\&.
.SS "void Draw::showOnlyChat ()\fC [signal]\fP"

.PP
Zeigt nur den \fBChat\fP\&. showOnlyChat 

.SH "Author"
.PP 
Generated automatically by Doxygen for Version from the source code\&.


draw.cpp
#include "draw.h"

Draw::Draw(QWidget *parent) : QWidget(parent)
{
    parentWindow = parent;

    win = false;
    ki = false;
    winChar = "";
    whichActionKI = 0;

    /* Bools initialization */
    showFps = true;

    /* Int initialization, Fps wird später die Fps errechnen und i wird der Zähler der Fps pro Sekunde sein. */
    fpsInt = 0;
    i = 0;

    /* Für das zeichnen zuständig, je schneller je öfter */
    timerUpdate = new QTimer(this);
    connect(timerUpdate, SIGNAL(timeout()), this, SLOT(update()));
    /* Das sind 60 FPS auf Mac, Besser schreibe statt 20 -> 16 rein für Windows. */
    if (getOSName() == "windows") {
        timerUpdate->start(15);
    } else {
        timerUpdate->start(20);
    }
    qsrand((uint)timerUpdate);

    /* Das sind die Sekunden, sie werden benutzt um die FPS auszurechnen. */
    seconds = new QTimer(this);
    connect(seconds, SIGNAL(timeout()), this, SLOT(fps()));
    seconds->start(1000);
}

void Draw::paintEvent(QPaintEvent *e) {
    e->accept();
    QPainter textPainter(this);

    if(!win) {
        background->drawBackground(&textPainter);
        UIinterface->drawInterface(&textPainter, character, enemy);
        character->drawChar(&textPainter, enemy);
        enemy->drawChar(&textPainter, character);
    } else {
        textPainter.setPen(QPen(Qt::red));
        textPainter.setBrush(QBrush(Qt::yellow));
        textPainter.setFont(QFont("Arial", 60, -1, false));
        textPainter.drawImage(QRect(350,100,100,100), winIcon.scaled(100,100));
        textPainter.drawText(QRect(0,0,800,600), "Spieler " + winChar + " gewann!", QTextOption(Qt::AlignCenter));
        textPainter.setFont(QFont("Arial", 12, -1, false));
        textPainter.drawText(QRect(0,500,800,100), "(Escape drücken um ins Startmenu zu kommen.)", QTextOption(Qt::AlignCenter));
    }

    i++;

    if (ki) {
        if (actionKI) {
            enemy->moveRight(false);
            enemy->moveLeft(false);
            whichActionKI = qrand() % ((5 + 1) - 0) + 0;

            if(enemy->enemyIsLeftRange(character) | enemy->enemyIsRightRange(character)) {
                whichActionKI = 5;
            }

            switch(whichActionKI) {
            case 0:
                break;
            case 1:
                if (!(enemy->getX() > 600)) {
                    enemy->moveRight(true);
                } else {
                    enemy->moveLeft(true);
                }
                break;
            case 2:
                if (!(enemy->getX() < 200)) {
                    enemy->moveLeft(true);
                } else {
                    enemy->moveRight(true);
                }
                break;
            case 3:
                if (!(enemy->getX() > 600)) {
                    enemy->moveRight(true);
                    enemy->jumpUp();
                } else {
                    enemy->moveLeft(true);
                    enemy->jumpUp();
                }
                break;
            case 4:
                if (!(enemy->getX() < 200)) {
                    enemy->moveLeft(true);
                    enemy->jumpUp();
                } else {
                    enemy->moveRight(true);
                    enemy->jumpUp();
                }
                break;
            case 5:
                if (enemy->enemyIsLeft(character)) {
                    enemy->moveLeft(true);
                    enemy->punch();
                } else if (enemy->enemyIsRight(character)) {
                    enemy->moveRight(true);
                    enemy->punch();
                }
                break;
            }
            actionKI = false;
        }
    }
}

int Draw::fps() {
    fpsInt = i;
    i = 0;
    return fpsInt;
}

void Draw::setFpsVisible(bool b) {
    showFps = b;
}

bool Draw::isFpsVisible() {
    return showFps;
}

void Draw::keyPressEvent(QKeyEvent *e) {
    switch (e->key()) {
    case Qt::Key_Q:
        character->punch();
        break;
    case Qt::Key_Escape:
        emit setCurrent(0);
        break;
    case Qt::Key_D:
        character->moveRight(true);
        break;
    case Qt::Key_A:
        character->moveLeft(true);
        break;
    case Qt::Key_S:
        character->setCrouch(true);
        break;
    case Qt::Key_W:
        character->jumpUp();
        break;
    case Qt::Key_Shift:
        if (!ki) {
            enemy->punch();
        }
        break;
    case Qt::Key_Left:
        if (!ki) {
            enemy->moveLeft(true);
        }
        break;
    case Qt::Key_Right:
        if (!ki) {
            enemy->moveRight(true);
        }
        break;
    case Qt::Key_Up:
        if (!ki) {
            enemy->jumpUp();
        }
        break;
    case Qt::Key_Down:
        if (!ki) {
            enemy->setCrouch(true);
        }
        break;
    case Qt::Key_Return:
        emit chatSignal();
        break;
    case Qt::Key_Y:
        emit showOnlyChat();
        break;
    }
}

void Draw::keyReleaseEvent(QKeyEvent *e) {
    switch (e->key()) {
    case Qt::Key_D:
        character->moveRight(false);
        break;
    case Qt::Key_A:
        character->moveLeft(false);
        break;
    case Qt::Key_S:
        character->setCrouch(false);
        break;
    case Qt::Key_Return:
        break;
    case Qt::Key_Y:
        emit hideOnlyChat();
        break;
    case Qt::Key_Left:
        if (!ki) {
            enemy->moveLeft(false);
        }
        break;
    case Qt::Key_Right:
        if (!ki) {
            enemy->moveRight(false);
        }
        break;
    case Qt::Key_Down:
        if (!ki) {
            enemy->setCrouch(false);
        }
        break;
    }
}

void Draw::load(int selChar, int selBackg, int selChar2) {
    character = new Character(selChar);
    background = new Background(selBackg);
    UIinterface = new UIOverlay();
    win = false;
    winChar = "";
    if(selChar2 == 0) {
        ki = true;
        kiActions = new QTimer(this);
        kiActions->setInterval(1500);
        connect(kiActions, SIGNAL(timeout()), this, SLOT(kiActionTrue()));
        kiActions->start();
        enemy = new Character(1, true);
    } else {
        ki = false;
        enemy = new Character(selChar2, true);
    }
    connect(enemy, SIGNAL(death()), this, SLOT(winC()));
    connect(character, SIGNAL(death()), this, SLOT(winE()));
}

QString Draw::getOSName()
{
    #if defined(Q_OS_ANDROID)
        return QLatin1String("android");
    #elif defined(Q_OS_BLACKBERRY)
        return QLatin1String("blackberry");
    #elif defined(Q_OS_IOS)
        return QLatin1String("ios");
    #elif defined(Q_OS_MAC)
        return QLatin1String("osx");
    #elif defined(Q_OS_WINCE)
        return QLatin1String("wince");
    #elif defined(Q_OS_WIN)
        return QLatin1String("windows");
    #elif defined(Q_OS_LINUX)
        return QLatin1String("linux");
    #elif defined(Q_OS_UNIX)
        return QLatin1String("unix");
    #else
        return QLatin1String("unknown");
    #endif
}

void Draw::winC() {
    QSound::play(":/fanfare.wav");
    winChar = character->getName();
    winIcon = character->getIcon();
    win = true;
}

void Draw::winE() {
    QSound::play(":/fanfare.wav");
    winChar = enemy->getName();
    winIcon = enemy->getIcon();
    win = true;
}

/**
 * @brief Draw::kiActionTrue
 *
 * Setzt die KI Action true sodass die KI beim nächsten Zug wieder eine Aktion machen soll.
 *
 */
void Draw::kiActionTrue() {
    actionKI = true;
}
